File: sdk/typescript/examples/06-retry.ts
/**
 * Retry & Dead Letter Queue
 */
import { FlashQ, Worker } from '../src';

const client = new FlashQ();

// Worker that fails jobs
let attempts = 0;
const worker = new Worker('flaky', async (job) => {
  attempts++;
  if (attempts < 3) {
    throw new Error(`Attempt ${attempts} failed`);
  }
  return { success: true };
}, { concurrency: 1 });

worker.on('failed', (job, err) => console.log('Failed:', err.message));
worker.on('completed', (job) => console.log('Success on attempt', attempts));

await worker.start();

// Push job with 3 retries
await client.push('flaky', { task: 'important' }, {
  max_attempts: 3,
  backoff: 100,
});

// Wait for retries
await new Promise(r => setTimeout(r, 3000));

// Check DLQ
const dlq = await client.getDlq('flaky');
console.log('Jobs in DLQ:', dlq.length);

await worker.stop();
await client.close();

File: sdk/python/examples/06_retry.py
"""
Example 06: Retry and DLQ

Demonstrates job retries and dead letter queue handling.
"""

import asyncio
from flashq import FlashQ, PushOptions


async def main():
    async with FlashQ() as client:
        # Push a job with max 3 attempts
        job_id = await client.push(
            "flaky-tasks",
            {"task": "might-fail"},
            PushOptions(max_attempts=3, backoff=500),
        )
        print(f"Pushed job: {job_id}")

        # Simulate failures
        for attempt in range(3):
            job = await client.pull("flaky-tasks", timeout=5000)
            if job:
                print(f"Attempt {job.attempts + 1}: Failing job {job.id}")
                await client.fail(job.id, f"Simulated failure #{attempt + 1}")

                # Wait for retry delay
                await asyncio.sleep(1)

        # Check DLQ
        dlq_jobs = await client.get_dlq("flaky-tasks")
        print(f"Jobs in DLQ: {len(dlq_jobs)}")

        for job in dlq_jobs:
            print(f"  - Job {job.id}: {job.error}")

        # Retry from DLQ
        if dlq_jobs:
            count = await client.retry_dlq("flaky-tasks")
            print(f"Retried {count} jobs from DLQ")


if __name__ == "__main__":
    asyncio.run(main())
